Flow: add loki.write component (#2563)
grafana/agent: docs/sources/flow/reference/components/loki.write.md

endpoint > http_client_config > oauth2 | [oauth2][] | Configure OAuth2 for authenticating to the endpoint. | no
endpoint > http_client_config > oauth2 > tls_config | [tls_config][] | Configure TLS settings for connecting to the endpoint. | no
endpoint > http_client_config > tls_config | [tls_config][] | Configure TLS settings for connecting to the endpoint. | no

The `>` symbol indicates deeper levels of nesting. For example, `endpoint >
http_client_config` refers to an `http_client_config` block defined inside an
`endpoint` block.

[endpoint]: #endpoint-block
[http_client_config]: #http_client_config-block
[basic_auth]: #basic_auth-block
[authorization]: #authorization-block
[oauth2]: #oauth2-block
[tls_config]: #tls_config-block

### endpoint block

The `endpoint` block describes a single location to send logs to. Multiple
`endpoint` blocks can be provided to send logs to multiple locations.

The following arguments are supported:

Name                  | Type       | Description                           | Default        | Required
--------------------- | -----------| ------------------------------------- | -------------- | --------

> `url`                 | `string`   | Full URL to send logs    to.          |                | yes

Reviewer: ```suggestion
`url`                 | `string`   | Full URL to send logs to.          |                | yes
```

`name`                | `string`   | Optional name to identify this endpoint with. |        | no
`batch_wait`          | `bool`     | Maximum amount of time to wait before sending a batch. | `"1s"` | no
`batch_size`          | `string`   | Maximum batch size of logs to accumulate before sending. | `"1MiB"` | no
`remote_timeout`      | `duration` | Timeout for requests made to the URL. | `"10s"`        | no
`tenant_id`           | `string`   | The tenant ID used by default to push logs. |  | no
`min_backoff_period`  | `duration` | Initial backoff time between retries. | `"500ms"`      | no
`max_backoff_period`  | `duration` | Maximum backoff time between retries. | `"5m"`         | no
`max_backoff_retries` | `int`      | Maximum number of retries.            | 10             | no

> If no `tenand_id` is provided, the component assumes that the Loki instance at

Reviewer: ```suggestion
If no `tenant_id` is provided, the component assumes that the Loki instance at
```

`endpoint` is running in single-tenant mode and no X-Scope-OrgID header is
sent.

When multiple `endpoint` blocks are provided, the `loki.write` component 
creates a client for each. Received log entries are fanned-out to these clients
in succession. That means that if one client is bottlenecked, it may impact

> the rest of them, too.

Reviewer: ```suggestion
the rest.
```

### authorization block

{{< docs/shared lookup="flow/reference/components/authorization-block.md" source="agent" >}}

### oauth2 block

{{< docs/shared lookup="flow/reference/components/oauth2-block.md" source="agent" >}}

### tls_config block

{{< docs/shared lookup="flow/reference/components/tls-config-block.md" source="agent" >}}

## Exported fields

The following fields are exported and can be referenced by other components:

Name | Type | Description
---- | ---- | -----------
`receiver` | `LogsReceiver` | A value that other components can use to send log entries to.

## Component health

`loki.write` is only reported as unhealthy if given an invalid
configuration.

## Debug information

`loki.write` does not expose any component-specific debug
information.

## Debug metrics
* `loki_write_encoded_bytes_total` (counter): Number of bytes encoded and ready to send.
* `loki_write_sent_bytes_total` (counter): Number of bytes sent.
* `loki_write_dropped_bytes_total` (counter): Number of bytes dropped because failed to be sent to the ingester after all retries.
* `loki_write_sent_entries_total` (counter): Number of log entries sent to the ingester.

> * `loki_write_dropped_entries_total` (counter): Number of log entries dropped because failed to be sent to the ingester after all retries.

Reviewer: ```suggestion
* `loki_write_dropped_entries_total` (counter): Number of log entries dropped because they failed to be sent to the ingester after all retries.
```

* `loki_write_request_duration_seconds` (histogram): Duration of sent requests.

> * `loki_write_batch_retries_total` (counter): Number of times batches has had to be retried.

Reviewer: ```suggestion
* `loki_write_batch_retries_total` (counter): Number of times batches have had to be retried.
```

* `loki_write_stream_lag_seconds` (gauge): Difference between current time and last batch timestamp for successful sends.

## Example

> This example creates a `loki.write` component that will send received entries
> to a local Loki instance without TLS:

Reviewer: ```suggestion
This example creates a `loki.write` component that sends received entries
to a local Loki instance without TLS:
```

`remote_timeout`      | `duration` | Timeout for requests made to the URL. | `"10s"`        | no
`tenant_id`           | `string`   | The tenant ID used by default to push logs. |  | no
`min_backoff_period`  | `duration` | Initial backoff time between retries. | `"500ms"`      | no
`max_backoff_period`  | `duration` | Maximum backoff time between retries. | `"5m"`         | no
`max_backoff_retries` | `int`      | Maximum number of retries.            | 10             | no

If no `tenand_id` is provided, the component assumes that the Loki instance at
`endpoint` is running in single-tenant mode and no X-Scope-OrgID header is
sent.

When multiple `endpoint` blocks are provided, the `loki.write` component 
creates a client for each. Received log entries are fanned-out to these clients
in succession. That means that if one client is bottlenecked, it may impact
the rest of them, too.

Endpoints can be named for easier identification in debug metrics using the
`name` argument. If the `name` argument isn't provided, a name is generated
based on a hash of the endpoint settings.

### http_client_config block

> The `http_client_config` configures settings used to connect to the
> remote_write server.

Reviewer 2: ```suggestion
The `http_client_config` configures settings used to connect to the
Loki API.
```

configuration.

## Debug information

`loki.write` does not expose any component-specific debug
information.

## Debug metrics
* `loki_write_encoded_bytes_total` (counter): Number of bytes encoded and ready to send.
* `loki_write_sent_bytes_total` (counter): Number of bytes sent.
* `loki_write_dropped_bytes_total` (counter): Number of bytes dropped because failed to be sent to the ingester after all retries.
* `loki_write_sent_entries_total` (counter): Number of log entries sent to the ingester.
* `loki_write_dropped_entries_total` (counter): Number of log entries dropped because failed to be sent to the ingester after all retries.
* `loki_write_request_duration_seconds` (histogram): Duration of sent requests.
* `loki_write_batch_retries_total` (counter): Number of times batches has had to be retried.
* `loki_write_stream_lag_seconds` (gauge): Difference between current time and last batch timestamp for successful sends.

## Example

This example creates a `loki.write` component that will send received entries
to a local Loki instance without TLS:

```river
loki.write "local" {
    client {
        url = "loki:3100"
        tls_config {
            insecure_skip_verify = true

>         }

Author: No it's not, the example just looked empty. I'll remove this.
